src/capture/capture_video.h

```diff
@@ -23,14 +23,152 @@
 #define DOSBOX_CAPTURE_VIDEO_H
 
 #include "render.h"
+#include "rwqueue.h"
 
-void capture_video_add_frame(const RenderedImage& image,
-                             const float frames_per_second);
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libswresample/swresample.h>
+#include <libswscale/swscale.h>
+}
 
-void capture_video_add_audio_data(const uint32_t sample_rate,
-                                  const uint32_t num_sample_frames,
-                                  const int16_t* sample_frames);
+class VideoEncoder {
+public:
+	virtual void capture_video_add_frame(const RenderedImage& image,
+	                                     const float frames_per_second) = 0;
 
-void capture_video_finalise();
+	virtual void capture_video_add_audio_data(const uint32_t sample_rate,
+	                                          const uint32_t num_sample_frames,
+	                                          const int16_t* sample_frames) = 0;
+
+	virtual void capture_video_finalise() = 0;
+
+	virtual ~VideoEncoder() = default;
+};
+
+class ZMBVEncoder : public VideoEncoder {
+public:
+	void capture_video_add_frame(const RenderedImage& image,
+	                             const float frames_per_second) override;
+
+	void capture_video_add_audio_data(const uint32_t sample_rate,
+	                                  const uint32_t num_sample_frames,
+	                                  const int16_t* sample_frames) override;
+
+	void capture_video_finalise() override;
+};
+
+struct FFmpegVideoEncoder {
+	RWQueue<RenderedImage> queue{256};
+	std::thread thread = {};
+
+	// FFmpeg pointers. Initialsed in main thread. Used by worker thread.
+	const AVCodec* codec          = nullptr;
+	AVCodecContext* codec_context = nullptr;
+	AVFrame* frame                = nullptr;
+	SwsContext* rescaler_contex   = nullptr;
+
+	// Accessed only in main thread, used to check if needs re-init
+	// If one of these changes, create a new file
+	PixelFormat pixel_format = {};
+	int frames_per_second    = 0;
+	uint16_t width           = 0;
+	uint16_t height          = 0;
+
+	// Synchronoization flags, guarded by mutex
+
+	// Main thread waits for worker to finish working.
+	bool is_working = false;
+
+	// Worker thread waits on main thread to initalise.
+	bool is_initalised = false;
+
+	// Init + free called by main thread only.
+	bool init(const uint16_t width, const uint16_t height,
+	          const PixelFormat pixel_format, const int frames_per_second);
+	void free();
+};
+
+struct FFmpegAudioEncoder {
```

Comment:
Let's not encode implementation details in class names. `MpegAudioEncoder` would be better.